### PR TITLE
Migrate to .NET Core v2

### PR DESCRIPTION
**Pull request by keyboardDrummer**
This PR is a modified version of [this similar one](https://github.com/dafny-lang/dafny/pull/778). The addition of this PR is that a forked version of Boogie is used as a submodule, with a change that makes Boogie easier to build so no manual steps are required to get Dafny to build after doing a `git clone --recurse-submodules`.

### Changes
- Convert all .csproj files from .NET framework to .NET Core.
- Include Boogie as a git submodule. The referenced Boogie is a modified version of v2.4.21 that does not reference the GitVersionTask package. Without this package it's much easier to build Boogie.

### Implementation details
- Remove reference to `Mono.Cecil` since it is no longer needed.
- Make some method calls in Dafny more explicit to resolve ambiguous references caused by changes in the standard library.
- Remove `Source/Dafny/cce.cs` in favor of [this Boogie version](https://github.com/boogie-org/boogie/blob/master/Source/CodeContractsExtender/cce.cs) which seems slightly better. This solves ambiguous references that occurred in the .NET Core build. It's not clear why didn't occur before.
- The `ModelViewer` package from Boogie, which is used by `DafnyServer` has not been ported to .NET Core, and I assume it can not be because it uses .NET framework libraries. It's unclear whether the `ModelViewer` package has a future in Boogie, since they want to get rid of their .NET framework support altogether. The files used from `ModelViewer` have been copied into `DafnyServer` to resolve the issue.
- Rewrite the `CompileTargetProgram` method in `Compiler-CSharp.cs` so it uses Roslyn's `Microsoft.CodeAnalysis.CSharp` to Compile C#, because the existing method of compiling with `System.CodeDom` throws a platform not supported exception for Ubuntu in .NET Core. See [this thread](https://github.com/dotnet/runtime/issues/18768) for more information.
- Update the lit configuration so it uses `dotnet run` to get a platform independent way to run, instead of using Mono to run `Dafny.exe`.

#### No more .NET Framework support
This PR adds support for the .NET Core framework but removes it for the .NET framework. Supporting both platforms at the same time would be difficult, since this PR contains source level changes. To support both platforms at the same time we'd have to duplicate some files and customize each platform to use the right file, but in that case I would suggest considering a fork for .NET Framework support.

### Testing
- Builds and runs on Windows using Visual studio or the dotnet CLI tool
- Builds and runs on MacOS using JetBrains Rider or the dotnet CLI tool

### CODE REVIEW

**Comment by RustanLeino** [comment]
These are two useful big changes, moving to .NET Core and incorporating Boogie as a git submodule. Thank you for your initiative and contributions!

I've tried to merge this with master and I'm having two problems.

One problem is building the parser and scanner, which uses the parser generator Coco/R (found in the `third_party` folder). Until recently, any change to `Dafny.atg` to had be accompanied with a matching change in `Parser.cs` and `Scanner.cs`, which were produced by manually starting Coco. This was changed in PR https://github.com/dafny-lang/dafny/pull/861, which adds Coco as a pre-build event. There are three differences in using this pre-build event with the current PR:

* The working directory (the output directory) of pre-build events is one level deeper, so there needs to be an additional `../` in the three paths in the pre-build event.
* The pre-build event seems to run _after_ the build system has determined which files to include in the build. So, if you manually delete `Parser.cs` and `Scanner.cs` and let the pre-build event rebuild these, then the compiler doesn't yet know about these file, so it generates some errors. If you build again, then this seems to work (but is probably using the prior version of `Parser.cs` and `Scanner.cs`). I don't know how to move this pre-build event earlier.
* When building using `msbuild Dafny.sln`, the build system automatically figures out that `Coco.exe` is a .NET executable, so it prefixes it with `mono` for platforms that need it. This is not happening with `dotnet build Dafny.sln`. One remedy would be to manually add `"mono"` to the pre-build event, but I suspect that would make the build not work on Windows.

The other problem I'm having is that `DafnyDriver` and `DafnyServer` output DLLs, not EXEs. As far as I can tell, the `.csproj` file specifies an `Exe`, but the build still throws in `/target:library` when calling `csc`, and thus you get `Dafny.dll` instead of `Dafny.exe`.

Any ideas of how to solve these problems?

**Comment by RustanLeino** [comment]
Oh, one more thing. The PR includes the `Test/dafny0/RuntimeTypeTests0` folder, but I think this is an error. This directory is generated when running the `RuntimeTypeTests0.dfy`, so it should be ignored, not checked into the repository.

**Comment by keyboardDrummer** [comment]
> The working directory (the output directory) of pre-build events is one level deeper, so there needs to be an additional ../ in the three paths in the pre-build event.

Should be solved.

> The pre-build event seems to run after the build system has determined which files to include in the build. So, if you manually delete Parser.cs and Scanner.cs and let the pre-build event rebuild these, then the compiler doesn't yet know about these file, so it generates some errors. If you build again, then this seems to work (but is probably using the prior version of Parser.cs and Scanner.cs). I don't know how to move this pre-build event earlier.

Solved using a different method for specifying a prebuild target.

> When building using msbuild Dafny.sln, the build system automatically figures out that Coco.exe is a .NET executable, so it prefixes it with mono for platforms that need it. This is not happening with dotnet build Dafny.sln. One remedy would be to manually add "mono" to the pre-build event, but I suspect that would make the build not work on Windows.

Solved by using `dotnet tool run` instead of Coco.exe so it's platform independent. This is similar to what Boogie already does [here](https://github.com/boogie-org/boogie/blob/master/Source/Core/Makefile#L12)

> The other problem I'm having is that DafnyDriver and DafnyServer output DLLs, not EXEs. As far as I can tell, the .csproj file specifies an Exe, but the build still throws in /target:library when calling csc, and thus you get Dafny.dll instead of Dafny.exe

On windows those two projects generate .exe files for me. How can I reproduce this problem?

**Comment by davidcok** [comment]
Remy - I first had the pre-build event build the parser, but then added the parser build into msbuild. So the msbuild should always run Coco and then the compile. Is this not the case? In any case, that is what the dotnet build needs to do.

- David

File: .github/workflows/msbuild.yml [hunk 1]
```diff
@@ -22,35 +22,26 @@ jobs:
...
         wget https://github.com/Z3Prover/z3/releases/download/z3-4.8.4/z3-4.8.4.d6df51951f4c-x64-ubuntu-14.04.zip
         unzip z3*.zip && rm *.zip
         cp -r z3* dafny/Binaries/z3
-    - name: Nuget Restore Dafny
-      run: nuget restore dafny/Source/Dafny.sln
+    - name: Setup dotnet
+      uses: actions/setup-dotnet@v1
+      with:
+        dotnet-version: '3.1.x' # SDK Version to use; x will use the latest version of the 3.1 channel
     - name: Build Dafny Parser
```
Comment (keyboardDrummer) [inline .github/workflows/msbuild.yml:40]: I think we can remove this since the dotnet build below here will also run Coco

**Comment by RustanLeino** [comment]
For the record, for anyone else going from an earlier commit of Dafny to this PR, here what I did that worked for me. Before you build (either inside an IDE like Rider or from the command-line with `msbuild Dafny.sln`), do

    git submodule update --init

from the root Dafny directory. This pulls in the Boogie submodule into the $DAFNY/Source/boogie folder. Once you've done this, you can start the Dafny build as usual.

(If for some reason you have to revert to a previous commit of Dafny, you'll need to run `msbuild Dafny.sln`, then manually delete the `bin` and `obj` folders below $DAFNY/Source, then run `nuget restore Dafny.sln`, and finally run `msbuild Dafny.sln` again.)

The new Dafny does not build `Dafny.exe` (or `DafnyServer.exe`), at least not on a Mac. Instead, it builds `Dafny.dll` (and `DafnyServer.dll`). That's apparently how .NET Core does it. When you previously would have run `mono Dafny.exe`, you will now instead run `dotnet Dafny.dll`. I adjusted the `dafny` script in the Binaries folder to do this.

@fabianhauser When we merge this PR and eventually make a Dafny binary release from it, then the VS Code plug-in will probably need to be changed as mentioned in my previous paragraph.

**Review by RustanLeino — CHANGES_REQUESTED** [review]
Things now work for me (on a Mac) and look good, except that the Dafny version number is missing when I run Dafny from the command line. How did it disappear?

**Comment by fabianhauser** [comment]
@RustanLeino To generate a deliverable self contained `.exe` you have to use `dotnet publish` with .NET Core (Also [dotnet deploying documentation: produce an executable](hhttps://docs.microsoft.com/en-us/dotnet/core/deploying/#produce-an-executable))

**Comment by davidcok** [comment]
I installed dotnet on my Mac and pulled this PR into a new local clone of Dafny. However, I only get problems when trying to build. I note that the instructions for building from source on Linux & Mac on https://github.com/dafny-lang/dafny/wiki/INSTALL have not been updated for the new dotnet configuration. I expect I'm missing some other installation requirements, but it is not clear what they are.

File: Binaries/dafny-server [hunk 2]
```diff
@@ -11,16 +9,26 @@ while [ -h "$SOURCE" ]; do
...
-DAFNYSERVER="$DAFNY_ROOT/DafnyServer.exe"
 
-if [[ ! -x "$MONO" ]]; then
-    echo "Error: Dafny requires Mono to run on non-Windows systems."
-    exit 1
+MY_OS=$(uname -s)
+if [ "${MY_OS:0:5}" == "MINGW" ] || [ "${MY_OS:0:6}" == "CYGWIN" ]; then
+    DAFNYSERVER_EXE_NAME="Dafny.exe"
+else
+    DAFNYSERVER_EXE_NAME="Dafny.dll"
```
Comment (keyboardDrummer) [inline Binaries/dafny-server:17]: Missing Server

File: Binaries/dafny-server [hunk 3]
```diff
@@ -11,16 +9,26 @@ while [ -h "$SOURCE" ]; do
...
 done
 DAFNY_ROOT="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
-DAFNYSERVER="$DAFNY_ROOT/DafnyServer.exe"
 
-if [[ ! -x "$MONO" ]]; then
-    echo "Error: Dafny requires Mono to run on non-Windows systems."
-    exit 1
+MY_OS=$(uname -s)
+if [ "${MY_OS:0:5}" == "MINGW" ] || [ "${MY_OS:0:6}" == "CYGWIN" ]; then
+    DAFNYSERVER_EXE_NAME="Dafny.exe"
```
Comment (keyboardDrummer) [inline Binaries/dafny-server:15]: Missing Server

File: Source/Dafny/Makefile [hunk 4]
```diff
@@ -1,20 +1,17 @@
...
-COCO = $(COCO_EXE_DIR)\Coco.exe
 
 # "all" depends on 2 files, really (Parser.cs and Scanner.cs), but they
 # are both generated in one go and I don't know a better way to tell
 # nmake that.  --KRML
 all: Parser.cs
 
 Parser.cs: $(FRAME_DIR)\scanner.frame $(FRAME_DIR)\parser.frame Dafny.atg
-	$(COCO) Dafny.atg -namespace Microsoft.Dafny -frames $(FRAME_DIR)
+	dotnet tool run Dafny.atg -namespace Microsoft.Dafny -frames $(FRAME_DIR)
```
Comment (keyboardDrummer) [inline Source/Dafny/Makefile:14]: Missing coco

File: Source/Dafny/Makefile.Linux [hunk 5]
```diff
@@ -3,25 +3,18 @@
...
-    COCO = mono $(COCO_EXE_DIR)/Coco.exe
-endif
-
 
 # "all" depends on 2 files, really (Parser.cs and Scanner.cs), but they
 # are both generated in one go and I don't know a better way to tell
 # nmake that.  --KRML
 default:
-	$(COCO) Dafny.atg -namespace Microsoft.Dafny -frames $(FRAME_DIR)
+	dotnet tool run Dafny.atg -namespace Microsoft.Dafny -frames $(FRAME_DIR)
```
Comment (keyboardDrummer) [inline Source/Dafny/Makefile.Linux:12]: Missing coco

File: Source/Dafny/Makefile.Linux [hunk 6]
```diff
@@ -3,25 +3,18 @@
...
 
 all: Parser.cs Scanner.cs
 
 Parser.cs Scanner.cs: $(FRAME_DIR)/scanner.frame $(FRAME_DIR)/parser.frame Dafny.atg
-	$(COCO) Dafny.atg -namespace Microsoft.Dafny -frames $(FRAME_DIR)
+	dotnet tool run Dafny.atg -namespace Microsoft.Dafny -frames $(FRAME_DIR)
```
Comment (keyboardDrummer) [inline Source/Dafny/Makefile.Linux:17]: Missing coco

File: Source/DafnyDriver/Properties/launchSettings.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (keyboardDrummer) [inline Source/DafnyDriver/Properties/launchSettings.json:1]: File should not be here

File: Source/Dafny/DafnyPipeline.csproj [hunk 8]
```diff
@@ -1,227 +1,35 @@
...
-    <Compile Include="Rewriter.cs" />
-    <Compile Include="SccGraph.cs" />
-    <Compile Include="Translator.cs" />
-    <Compile Include="..\version.cs" />
-  </ItemGroup>
-  <ItemGroup>
-    <Compile Include="cce.cs" />
-  </ItemGroup>
-  <ItemGroup>
-    <None Include="Dafny.atg" />
```
Comment (keyboardDrummer) [inline Source/Dafny/DafnyPipeline.csproj:186]: Does this need to be included on the right to trigger recompilation when it's changed?

**Comment by keyboardDrummer** [comment]
> Things now work for me (on a Mac) and look good, except that the Dafny version number is missing when I run Dafny from the command line. How did it disappear?

That was a bug. It's resolved now.

**Comment by keyboardDrummer** [comment]
> I installed dotnet on my Mac and pulled this PR into a new local clone of Dafny. However, I only get problems when trying to build. I note that the instructions for building from source on Linux & Mac at https://github.com/dafny-lang/dafny/blob/master/INSTALL.md and https://github.com/dafny-lang/dafny/wiki/INSTALL have not been updated for the new dotnet configuration. I expect I'm missing some other installation requirements, but it is not clear what they are.

I've updated the `INSTALL.md` file. Could you check if the updated instructions work for you?

**Comment by keyboardDrummer** [comment]
> @RustanLeino To generate a deliverable self contained `.exe` you have to use `dotnet publish` with .NET Core (Also see [dotnet deploying documentation: produce an executable](https://docs.microsoft.com/en-us/dotnet/core/deploying/#produce-an-executable))

@fabianhauser, I'm guessing that for the VSCode extension it will be easiest to run Dafny using `dotnet Dafny.dll`, as is mentioned [here](https://docs.microsoft.com/en-us/dotnet/core/deploying/#produce-a-cross-platform-binary), instead of using platform specific executables, right?

**Comment by davidcok** [comment]
The build (as in dotnet build Dafny.sln) does not build DafnyRuntime.jar,  causing tests that exercise the compile-to-java to fail. KRML and I suggest that the build create the DafnyRuntime.jar if java is installed and issue a warning message if java is not installed.

**Comment by davidcok** [comment]
INSTALL.md for Mac: step 5 mentions mono. Perhaps this should be dotnet?

Step 7 mentions installing java. Suggest including that optional step before the build of Dafny, so that the dafny build can build the DafnyRuntime.jar if desired. Should also say that java should be at least version 8.

**Comment by keyboardDrummer** [comment]
> The build (as in dotnet build Dafny.sln) does not build DafnyRuntime.jar, causing tests that exercise the compile-to-java to fail. KRML and I suggest that the build create the DafnyRuntime.jar if java is installed and issue a warning message if java is not installed.

I'm not sure what you mean. Is there a test that should be added? All the tests were passing right?

> Step 7 mentions installing java. Suggest including that optional step before the build of Dafny, so that the dafny build can build the DafnyRuntime.jar if desired. Should also say that java should be at least version 8.

Maybe something for a follow-up PR? I've tried only to change the things related to .NET Core.

**Comment by davidcok** [comment]
The build creates the DafnyRuntime for cs, but not for java. Consequently, lit tests will fail out of the box. It is preferable if Dafny.sln builds DafnyRuntime.jar, so that it is guaranteed up to date. This could be put off to a subsequent PR if you like. Otherwise this PR is sufficiently ready to merge.

**Comment by keyboardDrummer** [comment]
> The build creates the DafnyRuntime for cs, but not for java. Consequently, lit tests will fail out of the box. It is preferable if Dafny.sln builds DafnyRuntime.jar, so that it is guaranteed up to date.

This is also the existing behavior right? The .NET Framework build does not trigger building the Dafny runtime for Java from what I see. I'd prefer to keep changing that out of scope for this PR, since the PR is big enough.

**Comment by RustanLeino** [comment]
> > The build creates the DafnyRuntime for cs, but not for java. Consequently, lit tests will fail out of the box. It is preferable if Dafny.sln builds DafnyRuntime.jar, so that it is guaranteed up to date.
> 
> This is also the existing behavior right? The .NET Framework build does not trigger building the Dafny runtime for Java from what I see. I'd prefer to keep changing that out of scope for this PR, since the PR is big enough.

Correct and I agree.

**Review by RustanLeino — APPROVED** [review]

**Pull request merged**